File: pipeline/lib/podcasts.py
"""Podcast episodes picked for a hike, checked and shaped for the phone (#1683).

reference/podcast_episodes.json is the judgement - which episodes, on which
hikes - and this module is the gate between it and the bucket. Its README
says what a row holds; this says what a row must be before a phone sees it.

JUNK COSTS THE ROW, NEVER THE LIST, the posture client/src/lib/
suggestedHikesData.ts takes toward a document somebody else wrote. But the
somebody here is the maintainer, editing by hand, so a dropped row is never
quiet: `validate` returns every drop with its reason, export_podcasts.py
refuses to upload a list that dropped anything, and
tests/test_export_podcasts.py fails on the committed file for the same
reason. A typo becomes a red pull request, not an episode that never shows.

WHAT IS PUBLISHED, AND WHAT IS NOT. The id, title, show, length and the two
anchors. `note` and `reviewed` stay in the reference file: they are for the
person reviewing the diff, and a phone has no use for either.

NOTHING HERE CHECKS THAT A HIKE ID EXISTS. The suggested hikes are exported
by the vector publish into a release folder this script never reads, so an id
that matches no hike is published and matches nothing on the phone - a card
that does not appear, never a wrong one.
"""

from __future__ import annotations

import datetime as dt
import re
from dataclasses import dataclass, field

#: Spotify's base-62 ids are 22 characters. Reasoned from every episode link
#: Spotify prints (`open.spotify.com/episode/<22 chars>`), not from a spec
#: Spotify publishes - its Web API reference calls the id a "base-62
#: identifier" without stating the length. Checked because the id is spliced
#: into a URL the phone opens and a URI the phone sends to Spotify, and a
#: string that is not an id has no business in either.
SPOTIFY_ID_PATTERN = re.compile(r"^[0-9A-Za-z]{22}$")

#: The A.T. is about 2,200 miles from Springer. A range past this is a typo,
#: not a trail the app does not know about yet - the only mile axis the phone
#: holds is the A.T.'s (client/src/lib/hikes.ts, trailHasMileAxis).
MAX_AT_MILE = 2300.0

#: Every field a row may carry. Anything else is a misspelling of one of
#: these, and a misspelt `at_miles` would silently anchor nothing.
ROW_FIELDS = frozenset({"spotify_id", "title", "show", "minutes", "hikes", "at_miles", "reviewed", "note"})


@dataclass(frozen=True)
class Episode:
    spotify_id: str
    title: str
    show: str
    minutes: int | None
    hikes: tuple[str, ...]
    at_miles: tuple[tuple[float, float], ...]


@dataclass
class Validation:
    episodes: list[Episode] = field(default_factory=list)
    #: (row label, reason) for every row that did not make it.
    dropped: list[tuple[str, str]] = field(default_factory=list)


def _text(value: object) -> str | None:
    return value.strip() if isinstance(value, str) and value.strip() else None


def _minutes(value: object) -> tuple[int | None, str | None]:
    if value is None:
        return None, None
    # bool is an int in Python, and `true` is not a length.
    if isinstance(value, bool) or not isinstance(value, int) or value <= 0:
        return None, f"minutes must be a positive whole number or absent, not {value!r}"
    return value, None


def _hikes(value: object) -> tuple[tuple[str, ...], str | None]:
    if value is None:
        return (), None
    if not isinstance(value, list) or not all(_text(item) for item in value):
        return (), f"hikes must be a list of hike ids, not {value!r}"
    return tuple(item.strip() for item in value), None


def _ranges(value: object) -> tuple[tuple[tuple[float, float], ...], str | None]:
    if value is None:
        return (), None
    if not isinstance(value, list):
        return (), f"at_miles must be a list of [start, end] pairs, not {value!r}"
    ranges = []
    for pair in value:
        if (
            not isinstance(pair, list)
            or len(pair) != 2
            or not all(isinstance(end, int | float) and not isinstance(end, bool) for end in pair)
        ):
            return (), f"each at_miles entry must be [start, end], not {pair!r}"
        start, end = float(pair[0]), float(pair[1])
        if not 0 <= start < end <= MAX_AT_MILE:
            return (), f"at_miles {pair!r} must run forward, inside 0 to {MAX_AT_MILE:g}"
        ranges.append((start, end))
    return tuple(ranges), None


def _reviewed(value: object) -> str | None:
    if not isinstance(value, str):
        return "reviewed must be a date, YYYY-MM-DD"
    try:
        dt.date.fromisoformat(value)
    except ValueError:
        return f"reviewed {value!r} is not a YYYY-MM-DD date"
    return None


def validate(rows: list[object]) -> Validation:
    """Every row that can be published, and why each other one cannot."""
    result = Validation()
    seen: set[str] = set()
    for at, row in enumerate(rows):
        label = f"row {at}"
        if not isinstance(row, dict):
            result.dropped.append((label, "not an object"))
            continue
        spotify_id = row.get("spotify_id")
        if isinstance(spotify_id, str):
            label = f"row {at} ({spotify_id})"

        unknown = sorted(set(row) - ROW_FIELDS)
        if unknown:
            result.dropped.append((label, f"unknown field(s) {', '.join(unknown)}"))
            continue
        if not isinstance(spotify_id, str) or not SPOTIFY_ID_PATTERN.match(spotify_id):
            result.dropped.append((label, "spotify_id must be the 22 characters after /episode/ in its link"))
            continue
        if spotify_id in seen:
            result.dropped.append((label, "the same episode is listed twice; merge the two rows"))
            continue
        title, show = _text(row.get("title")), _text(row.get("show"))
        if title is None or show is None:
            result.dropped.append((label, "title and show are both required"))
            continue
        hikes: tuple[str, ...] = ()
        at_miles: tuple[tuple[float, float], ...] = ()
        minutes, why = _minutes(row.get("minutes"))
        if why is None:
            hikes, why = _hikes(row.get("hikes"))
        if why is None:
            at_miles, why = _ranges(row.get("at_miles"))
        if why is None:
            why = _reviewed(row.get("reviewed"))
        if why is None and not hikes and not at_miles:
            why = "needs at least one of hikes or at_miles, or it shows nowhere"
        if why is not None:
            result.dropped.append((label, why))
            continue

        seen.add(spotify_id)
        result.episodes.append(Episode(spotify_id, title, show, minutes, hikes, at_miles))
    return result


def as_published(episode: Episode) -> dict:
    """The phone's copy of a row. `minutes` is left out rather than nulled
    when nobody gave one, the same absent-means-unknown rule the rest of the
    published data keeps."""
    record: dict = {
        "spotify_id": episode.spotify_id,
        "title": episode.title,
        "show": episode.show,
        "hikes": list(episode.hikes),
        "at_miles": [list(pair) for pair in episode.at_miles],
    }
    if episode.minutes is not None:
        record["minutes"] = episode.minutes
    return record

File: client/src/lib/podcasts.ts
// Podcast episodes picked for a hike (#1683 - Offer podcast episodes picked
// for the hike, with a one-tap Spotify save and an in-app player).
//
// WHAT THIS READS. `podcasts/episodes.json`, written by
// pipeline/export_podcasts.py from the reviewed
// pipeline/reference/podcast_episodes.json. Every row is somebody's editorial
// call; nothing here searches Spotify or ranks anything.
//
// AT THE BUCKET ROOT, NOT IN THE PINNED RELEASE (lib/dataRelease.ts's
// ROOT_SCOPED_PREFIXES). The maintainer chose a live list, 2026-09-26: a new
// episode reaches this phone on its next fetch, with no app release and no
// DATA_RELEASE bump. So the document can change under a running build, which
// is why the parse below takes nothing on trust.
//
// NO POSITION LEAVES THE PHONE. The whole list is fetched and matched here,
// against a hike id or the miles of the day a hiker planned. Spotify hears
// about an episode only when the hiker taps it (chrome/PodcastCard.tsx).
//
// NEVER FATAL, lib/mapSheets.ts's posture through the same cache: an
// unreachable bucket, a 404 from an environment nobody has published to, or
// a document this build cannot read all yield the kept copy or nothing, and
// nothing is what every screen rendered before this existed.

import { DATA_CONFIGURED, dataUrl } from './config'
import { recallPublished, rememberPublished } from './conditionsCache'

/** The key pipeline/export_podcasts.py writes. Must match exactly:
 *  pipeline/tests/test_published_key_contract.py holds the two spellings
 *  together, as it does for every other key the app fetches. */
export const PODCAST_EPISODES_KEY = 'podcasts/episodes.json'

export interface PodcastEpisode {
  /** Spotify's 22-character episode id - the tail of its open.spotify.com
   *  link. */
  readonly spotifyId: string
  readonly title: string
  readonly show: string
  /** Whole minutes, or undefined where nobody gave a length. Undefined
   *  prints no length - never a zero. */
  readonly minutes?: number
  /** Suggested-hike ids (lib/suggestedHikes.ts's `SuggestedHike.id`) the
   *  hike detail screen shows it on. */
  readonly hikes: readonly string[]
  /** A.T. mile ranges, [low, high], counted from Springer. Typed by hand from
   *  a guidebook (the reference file's README says so), so loose by a mile or
   *  two against this phone's own axis - which an overlap test absorbs and a
   *  podcast can afford. */
  readonly atMiles: readonly (readonly [number, number])[]
}

export const NO_PODCAST_EPISODES: readonly PodcastEpisode[] = []

/** The same rule pipeline/lib/podcasts.py enforces, checked again here
 *  because the id is spliced into a URL this phone loads in a frame and a
 *  URI it sends to Spotify. A document that went through the pipeline always
 *  passes; this is for the one that did not. */
const SPOTIFY_ID = /^[0-9A-Za-z]{22}$/

function text(value: unknown): string | null {
  return typeof value === 'string' && value.trim() !== '' ? value.trim() : null
}

function milesPair(value: unknown): readonly [number, number] | null {
  if (!Array.isArray(value) || value.length !== 2) return null
  const [low, high] = value as unknown[]
  if (typeof low !== 'number' || typeof high !== 'number') return null
  if (!Number.isFinite(low) || !Number.isFinite(high) || low < 0 || high <= low)
    return null
  return [low, high]
}

/** One published row, or null. Junk costs the row, never the list. */
function parseEpisode(value: unknown): PodcastEpisode | null {
  if (typeof value !== 'object' || value === null) return null
  const record = value as Record<string, unknown>
  const spotifyId = typeof record.spotify_id === 'string' ? record.spotify_id : ''
  const title = text(record.title)
  const show = text(record.show)
  if (!SPOTIFY_ID.test(spotifyId) || title === null || show === null) return null

  const hikes = Array.isArray(record.hikes)
    ? record.hikes.map(text).filter((id): id is string => id !== null)
    : []
  const atMiles = Array.isArray(record.at_miles)
    ? record.at_miles.map(milesPair).filter((pair) => pair !== null)
    : []
  // An episode anchored to nothing would show nowhere; keeping it would only
  // make a count somewhere disagree with what a hiker can find.
  if (hikes.length === 0 && atMiles.length === 0) return null

  const minutes =
    typeof record.minutes === 'number' &&
    Number.isInteger(record.minutes) &&
    record.minutes > 0
      ? record.minutes
      : undefined
  return {
    spotifyId,
    title,
    show,
    hikes,
    atMiles,
    ...(minutes === undefined ? {} : { minutes }),
  }
}

/** The published document's episodes, or null when it is not the shape this
 *  build reads at all. An empty list is a real answer: nobody has picked any. */
export function parsePodcastEpisodes(document: unknown): PodcastEpisode[] | null {
  if (typeof document !== 'object' || document === null) return null
  const episodes = (document as Record<string, unknown>).episodes
  if (!Array.isArray(episodes)) return null
  const seen = new Set<string>()
  const parsed: PodcastEpisode[] = []
  for (const value of episodes) {
    const episode = parseEpisode(value)
    if (episode === null || seen.has(episode.spotifyId)) continue
    seen.add(episode.spotifyId)
    parsed.push(episode)
  }
  return parsed
}

/** The episodes picked for one suggested hike, in the list's own order. */
export function episodesForHike(
  episodes: readonly PodcastEpisode[],
  hikeId: string,
): readonly PodcastEpisode[] {
  const found = episodes.filter((episode) => episode.hikes.includes(hikeId))
  return found.length === 0 ? NO_PODCAST_EPISODES : found
}

/**
 * The episodes whose A.T. range overlaps the stretch [low, high].
 *
 * Overlap rather than containment: a day from mi 486 to 497 inside an
 * episode picked for mi 480-512 is the case this exists for, and so is a day
 * that only clips its end. Touching ends count - a range that stops at the
 * shelter a day starts from is about the ground either side of it.
 */
export function episodesForMiles(
  episodes: readonly PodcastEpisode[],
  low: number,
  high: number,
): readonly PodcastEpisode[] {
  const found = episodes.filter((episode) =>
    episode.atMiles.some(([start, end]) => start <= high && end >= low),
  )
  return found.length === 0 ? NO_PODCAST_EPISODES : found
}

/** Where the episode opens in Spotify - its app, where one is installed and
 *  the link is handed to the system, or the web player. */
export function spotifyEpisodeUrl(episode: PodcastEpisode): string {
  return `https://open.spotify.com/episode/${episode.spotifyId}`
}

/** Spotify's own embedded player for the episode (chrome/PodcastCard.tsx
 *  loads it only once a hiker taps Play). */
export function spotifyEmbedUrl(episode: PodcastEpisode): string {
  return `https://open.spotify.com/embed/episode/${episode.spotifyId}`
}

/** `48 min`, `1 h 12 min`, or null where nobody gave a length. */
export function formatMinutes(minutes: number | undefined): string | null {
  if (minutes === undefined) return null
  if (minutes < 60) return `${minutes} min`
  const hours = Math.floor(minutes / 60)
  const rest = minutes % 60
  return rest === 0 ? `${hours} h` : `${hours} h ${rest} min`
}

/**
 * The list, from the bucket when there is signal and from the copy this
 * phone kept when there is not.
 *
 * A 404 is ordinary - an environment nobody has published a list to - and
 * falls back to the kept copy rather than clearing it: a bucket that stopped
 * serving the object is not evidence the episodes were withdrawn.
 */
export async function fetchPodcastEpisodes(
  online: boolean,
  signal?: AbortSignal,
): Promise<PodcastEpisode[] | null> {
  // The kept copy first and unconditionally, like useSuggestedHikes' recall:
  // a copy this phone holds is an answer whether or not this build could
  // fetch a fresh one.
  if (!online) return recalled()
  if (!DATA_CONFIGURED) return null
  try {
    const response = await fetch(dataUrl(PODCAST_EPISODES_KEY), { signal })
    if (!response.ok) return recalled()
    const document: unknown = await response.json()
    const parsed = parsePodcastEpisodes(document)
    // Kept only when it parsed, so a shape this build refuses never
    // overwrites the last one it could read.
    if (parsed !== null)
      void rememberPublished(PODCAST_EPISODES_KEY, document as Record<string, unknown>)
    return parsed ?? (await recalled())
  } catch {
    return recalled()
  }
}

async function recalled(): Promise<PodcastEpisode[] | null> {
  const cached = await recallPublished(PODCAST_EPISODES_KEY)
  if (cached === null) return null
  return parsePodcastEpisodes(cached.document)
}
